### DISCUSSION TITLE
Guide to migrate configs to support Colors index reference

### DISCUSSION THREAD

**Discussion by 7iomka** [mantinedev/mantine#2052] [Q&A]
Hi.
In v 5.1 you give us supporting [Colors index reference](https://mantine.dev/theming/colors/#colors-index-reference) feature.
Thanks.
But I have one big problem - how to migrate configs to support this feature?
Our config have fallback to primary color, we also use our custom palette.
When you use color index reference, you have 2 errors
- typescript: `'"green.2"' is not assignable to type '"primary" | "secondary" | "success" | "danger" | "dark" | "grape" | "slate" | "zin..... Did you mean '"green"`
- configs like described below breaks code, as a result of getting props of undefined result (theme.colors[colorProp] cannot be found)

This is an example of config for provider:

```tsx
const buttonStyles = (theme: MantineTheme, params: ButtonStylesParams) => {
  let textColor;
  let hoverStyles = {};
  let borderColor;

  switch (params.variant) {
    case 'filled': {
      const colorProp = params.color || theme.primaryColor;
      if (colorProp === 'secondary') {
        textColor = '#fff'; // secondary color by design force white text color
      } else {
        textColor = getReadableColor({
          color: theme.colors[colorProp][theme.primaryShade as number],
        });
      }

      hoverStyles = {
        backgroundColor: theme.colors[params.color || theme.primaryColor][6],
      };
      break;
    }

    case 'outline':
      textColor = params.color || 'inherit';
      borderColor = params.color || 'currentColor';
      hoverStyles = {
        borderColor: theme.colors[params.color || theme.primaryColor][theme.primaryShade as number],
        backgroundColor: theme.fn.rgba(theme.colors[params.color || theme.primaryColor][2], 0.7),
      };
      break;
    case 'subtle':
      textColor = 'inherit';
      hoverStyles = {
        backgroundColor: theme.colors.gray[1],
      };
      break;

    default:
      break;
  }

  return {
    root: {
      fontWeight: 400,
      borderRadius: params.radius || theme.radius.xl,
      color: textColor,
      borderColor,
      '&:not(:disabled):hover': hoverStyles,
      '&[data-disabled]': {
        backgroundColor: theme.colorScheme === 'dark' ? theme.colors.dark[4] : theme.colors.gray[4],
        color: theme.colorScheme === 'dark' ? theme.colors.dark[6] : theme.colors.gray[6],
      },
    },
    leftIcon: {
      marginRight: 8,
    },
    rightIcon: {
      marginLeft: 8,
    },
  };
};

// ....

Button: {
  styles: buttonStyles,
},

```

**Comment by rtivital** [comment]
By default, `MantineColor` type includes `(string & {})`, so if you do not redefine colors, it should work just fine. In case you do redine colors type then you can either append `| (string & {})` to your type, or generate type that will include every possible combination using interpolation:

```tsx
type ColorsName = 'blue' | 'pink' | 'red' | 'orange'; // ... other colors
type Shades = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;
type Colors = `${ColorsName}.${Shades}`
```

https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gGzgJwM4DkCGBbaBeKAcgCMEBXCQqAHyLAEsA7AaytsOQgBM2iVNGAc0oBuKAHpxUAHSyocYAAsIyKAGNEKVAChQkKAGVFmLhFRQCABhpQAjDYBMNgMw2ALDYCsNgGw2A7DYAHDYAnCK64NDwSGgWUAAGACQA3jFaWLgAvtKpRiZmWQlAA

Not sure about performance in later method, I use `(string & {})` in my applications.

**Comment by 7iomka** [comment]
Ok, thanks, that is enough for typings.
But for configuration how I need to refactor my color access code like `color: theme.colors[params.color || theme.primaryColor][theme.primaryShade as number]`? Many times I used this approach, but now seems we need some utility to easy handle this like
theme.fn.themeColor(color, shade?), where color under the hood can be prepared to support colors index reference

**Reply by rtivital** [reply]
Mantine components use `theme.fn.themeColor`, you can use source code as reference, I cannot advise much for strict mode though as I do not use it in any of my projects.

**Reply by 7iomka** [reply]
> Mantine components use `theme.fn.themeColor`, you can use source code as reference, I cannot advise much for strict mode though as I do not use it in any of my projects.

It is not mention on the docs, but is available. Why? Also I noticed that you migrate from use it to `theme.fn.variant`.

I write myself something like this
```tsx
type Shade = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;
interface MantinePrimaryShade {
  light: Shade;
  dark: Shade;
}
export const getThemeColor = ({
  theme,
  color = theme.primaryColor,
  shade = theme.primaryShade,
}: {
  theme: MantineTheme;
  color?: MantineColor;
  shade?: Shade | MantinePrimaryShade;
}) => {
  const colorProp = color;
  const colorArr = colorProp.split('.');
  const colorValue = colorArr[0];
  let shadeValue = shade;
  if (colorArr.length > 1) {
    shadeValue = Number(colorArr[1]) as Shade;
  }
  return theme.colors[colorValue][shadeValue as number];
};
```

I also not use struct mode.

**Reply by rtivital** [reply]
It is not listed in documentation website because its api is not stable. I would not recommend using undocumented features as they may change across releases without warning as they considered to be an internal tool.
